File: examples/youtube-info.lua
--
-- Echos back any links to youtube videos with the video's title, duration, and
-- thumbnail
--
-- Requires:  wget, jshon
--

local prefix = os.getenv('PREFIX') or "examples/"
local worker = prefix .. "youtube-info-worker.sh"

local message_fmt = [[
<table>
    <tr>
        <td valign="middle">
            <img src='https://www.youtube.com/yt/brand/media/image/YouTube-icon-full_color.png' height="25" />
        </td>
        <td align="center" valign="middle">
            <a href="http://youtu.be/%s">%s (%s)</a>
        </td>
    </tr>
    <tr>
        <td></td>
        <td align="center">
            <a href="http://youtu.be/%s"><img src="%s" width="250" /></a>
        </td>
    </tr>
</table>
]]

piepan.On('connect', function()
  print ("youtube-info loaded!")
end)

piepan.On('message', function(e)
  if e.Sender == nil then
    return
  end
  local patterns = {
    "https?://www%.youtube%.com/watch%?v=([%d%a_%-]+)",
    "https?://youtube%.com/watch%?v=([%d%a_%-]+)",
    "https?://youtu.be/([%d%a_%-]+)",
    "https?://youtube.com/v/([%d%a_%-]+)",
    "https?://www.youtube.com/v/([%d%a_%-]+)"
  }
  for _,pattern in ipairs(patterns) do
    local video_id = string.match(e.Message, pattern)
    if video_id ~= nil and string.len(video_id) < 20 then
      piepan.Process.New(function (success, data)
        if not success then
          return
        end

        local id, title, duration, thumbnail = string.match(data, "([^\r\n]+)\r?\n([^\r\n]+)\r?\n([^\r\n]+)\r?\n([^\r\n]+)\r?\n")
        duration = string.format("%d:%02d", duration / 60, duration % 60)
        local message = string.format(message_fmt, id, title, duration,
          id, thumbnail)
        piepan.Self.Channel().Send(message, false)
      end, worker, video_id)
      return
    end
  end
end)

File: examples/youtube-info.js
/*
 * Echos back any links to youtube videos with the video's title, duration,
 * and thumbnail.
 *
 * Requires:  wget, jshon
 */

(function() {

var prefix = ENV['PREFIX'] || 'examples/';
var worker = prefix + 'youtube-info-worker.sh';

var message_template = _.template('\
<table>\
    <tr>\
        <td valign="middle">\
            <img src="https://www.youtube.com/yt/brand/media/image/YouTube-icon-full_color.png" height="25" />\
        </td>\
        <td align="center" valign="middle">\
            <a href="http://youtu.be/<%= id %>"><%= title %> (<%= duration %>)</a>\
        </td>\
    </tr>\
    <tr>\
        <td></td>\
        <td align="center">\
            <a href="http://youtu.be/<%= id %>"><img src="<%= thumbnail %>" width="250" /></a>\
        </td>\
    </tr>\
</table>');

piepan.On('connect', function() {
  console.log('youtube-info loaded!')
});

piepan.On('message', function(e) {
  if (e.Sender == null) {
    return;
  }
  var patterns = [
    /https?:\/\/www\.youtube\.com\/watch\?v=([\w-]+)/,
    /https?:\/\/youtube\.com\/watch\?v=([\w-]+)/,
    /https?:\/\/youtu.be\/([\w-]+)/,
    /https?:\/\/youtube.com\/v\/([\w-]+)/,
    /https?:\/\/www.youtube.com\/v\/([\w-]+)/
  ];
  for (var i = 0; i < patterns.length; i++) {
    var pattern = patterns[i];
    var matches = e.Message.match(pattern);
    if (!matches) {
      continue;
    }
    var video_id = matches[1];
    if (video_id.length >= 20) {
      continue;
    }

    piepan.Process.New(function (success, data) {
      if (!success) {
        return;
      }

      var matches = data.match(/([^\r\n]+)\r?\n([^\r\n]+)\r?\n([^\r\n]+)\r?\n([^\r\n]+)\r?\n/);

      var minutes = (matches[3] / 60).toFixed(0).toString();
      var seconds = (matches[3] % 60).toFixed(0);
      var duration = minutes + ":";
      if (seconds < 10) {
        duration += "0";
      }
      duration += seconds;

      var message = message_template({
        id: matches[1],
        title: matches[2],
        duration: duration,
        thumbnail: matches[4],
      });
      piepan.Self.Channel().Send(message, false);
    }, worker, video_id);
    return;
  }
});

})();
